### PR TITLE
Fix a crash in the ImplicitNullChecks pass.

### PR DESCRIPTION
**Pull request by vargaz**
Re: https://github.com/dotnet/runtime/pull/46817#issuecomment-760542379

### CODE REVIEW

File: llvm/lib/CodeGen/ImplicitNullChecks.cpp [hunk 1]
```diff
@@ -346,6 +346,8 @@ ImplicitNullChecks::areMemoryOpsAliased(const MachineInstr &MI,
           return AR_MayAlias;
         continue;
       }
+      if (MMO2->getValue() == NULL)
```
Comment (imhameed) [inline llvm/lib/CodeGen/ImplicitNullChecks.cpp]: Superficial: `nullptr` instead of `NULL`

**Review by imhameed — APPROVED** [review]

File: llvm/lib/CodeGen/ImplicitNullChecks.cpp [hunk 2]
```diff
@@ -346,6 +346,8 @@ ImplicitNullChecks::areMemoryOpsAliased(const MachineInstr &MI,
           return AR_MayAlias;
         continue;
       }
+      if (MMO2->getValue() == nullptr)
+        continue;
```
Comment (imhameed) [inline llvm/lib/CodeGen/ImplicitNullChecks.cpp]: Spent a bit of time reading more closely. This case is hit if any the previous instructions in the basic block have a memory operand that has neither a `PseudoSourceValue` nor a `Value`. Maybe it'd be safest, for now, to assume that PrevMI and MI operate on aliased memory (e.g. just `return AR_MayAlias`)

**Comment by imhameed** [comment]
Here are two shrunken test cases. Build with `-march=x86-64 -mcpu=generic -enable-implicit-null-checks -filetype=obj`.

This one successfully compiles (https://godbolt.org/z/dvrqG8):
```llvm
@aotconst_sflda_22 = internal unnamed_addr global i64* null, align 8
@aotconst_sflda_23 = internal unnamed_addr global i64* null, align 8

define void @testfunc(i32 %branchguy, i64 * %dst) {
  %1 = icmp eq i32 %branchguy, 1

  %isnull = icmp eq i64 * %dst, null
  br i1 %isnull, label %EX_BB0, label %NOEX_BB0, !make.implicit !0

NOEX_BB0:
  %dst1 = bitcast i64 * %dst to i32 *
  store i32 0, i32 *%dst1, align 4

  %.val = load i64**, i64*** bitcast (i64** @aotconst_sflda_22 to i64***), align 8
  %.val4 = load i64**, i64*** bitcast (i64** @aotconst_sflda_23 to i64***), align 8
  %.in = select i1 %1, i64** %.val, i64** %.val4
  %2 = load i64*, i64** %.in, align 8

  %is_2_null = icmp eq i64 * %2, null
  br i1 %is_2_null, label %EX_BB1, label %NOEX_BB1, !make.implicit !0

EX_BB0:
  ret void

NOEX_BB1:
  store i32 0, i32 * %dst1, align 4
  ret void

EX_BB1:
  ret void
}

!0 = !{}
```

This one fails (https://godbolt.org/z/8sq6hd):
```llvm
@aotconst_sflda_22 = internal unnamed_addr global i64* null, align 8
@aotconst_sflda_23 = internal unnamed_addr global i64* null, align 8

define void @testfunc(i32 %branchguy, i64 * %dst) {
  %1 = icmp eq i32 %branchguy, 1

  %.val = load i64**, i64*** bitcast (i64** @aotconst_sflda_22 to i64***), align 8
  %.val4 = load i64**, i64*** bitcast (i64** @aotconst_sflda_23 to i64***), align 8
  %.in = select i1 %1, i64** %.val, i64** %.val4
  %2 = load i64*, i64** %.in, align 8

  %isnull = icmp eq i64 * %dst, null
  br i1 %isnull, label %EX_BB0, label %NOEX_BB0, !make.implicit !0

NOEX_BB0:
  %dst1 = bitcast i64 * %dst to i32 *
  store i32 0, i32 *%dst1, align 4

  %is_2_null = icmp eq i64 * %2, null
  br i1 %is_2_null, label %EX_BB1, label %NOEX_BB1, !make.implicit !0

EX_BB0:
  ret void

NOEX_BB1:
  store i32 0, i32 * %dst1, align 4
  ret void

EX_BB1:
  ret void
}

!0 = !{}
```

**Comment by vargaz** [comment]
No idea whenever this is fixed in llvm master, don't have a master build right now.

**Pull request merged**